### PR TITLE
Support masking for HopSkipJump attack

### LINKED ISSUE #624: Implement support for attack mask in HopSkipJump
A mask defines which of the input features are allowed to be modified by an attack. `art.attacks.evasion.HopSkipJump` 
 should provide support for masks similar to `mask` in `art.attacks.evasion.ProjectedGradientDescent` but accounting for the difference in gradient calculation.

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
To support the masking option for the HopSkipJump attack.

Fixes #624

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/hop_skip_jump.py [hunk 1]
```diff
@@ -140,7 +150,14 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
         x_adv_init = kwargs.get("x_adv_init")
 
         if x_adv_init is not None:
+            # Add mask param to the x_adv_init
+            if mask is not None:
+                for i in range(x.shape[0]):
+                    x_adv_init[i] = x_adv_init[i] * mask[i] + x[i] * (1 - mask[i])
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:156]: Is this loop necessary? Could it be replace by a vectorised operation?

File: art/attacks/evasion/hop_skip_jump.py [hunk 2]
```diff
@@ -188,8 +234,42 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+        mask = kwargs.get("mask")
+
+        if mask is not None:
+            if classifier_mixin:
+                # Ensure the mask is broadcastable
+                if len(mask.shape) > len(x.shape) or mask.shape != x.shape[-len(mask.shape) :]:
+                    raise ValueError("Mask shape must be broadcastable to input shape.")
+
+            else:
+                raise ValueError("Mask is only supported for classification.")
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:259]: I think the mask should also work for tasks other than classification without any additional modifications, or?

File: art/attacks/evasion/hop_skip_jump.py [hunk 3]
```diff
@@ -158,25 +175,54 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+                        x=val,
+                        y=y[ind],
+                        y_p=preds[ind],
+                        init_pred=init_preds[ind],
+                        adv_init=x_adv_init[ind],
+                        mask=None,
+                        clip_min=clip_min,
+                        clip_max=clip_max,
+                    )
+
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:201]: Would it be possible to change the value of `mask` so that we only need one function call? E.g. make mask a list of None if no mask is provided?

File: art/attacks/evasion/hop_skip_jump.py [hunk 4]
```diff
@@ -158,25 +175,54 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+                    x_adv[ind] = self._perturb(
+                        x=val,
+                        y=-1,
+                        y_p=preds[ind],
+                        init_pred=init_preds[ind],
+                        adv_init=x_adv_init[ind],
+                        mask=None,
+                        clip_min=clip_min,
+                        clip_max=clip_max,
+                    )
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:225]: Would it be possible to change the value of `mask` so that we only need one function call? E.g. make mask a list of None if no mask is provided?

File: art/attacks/evasion/hop_skip_jump.py [hunk 5]
```diff
@@ -188,8 +234,42 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
 
         return x_adv
 
+    @staticmethod
+    def _get_mask(x: np.ndarray, classifier_mixin: bool = True, **kwargs) -> np.ndarray:
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:238]:
```suggestion
    def _get_mask(x: np.ndarray, classifier_mixin: bool = True, **kwargs) -> List[Optional[np.ndarray]]:
```

File: art/attacks/evasion/hop_skip_jump.py [hunk 6]
```diff
@@ -127,6 +127,16 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
 
+        # Get the mask
+        mask = self._get_mask(x, **kwargs)
+
+        # Process mask as for each example
+        if mask is not None:
+            if len(mask.shape) == len(x.shape):
+                mask = mask.astype(ART_NUMPY_DTYPE)
+            else:
+                mask = np.array([mask.astype(ART_NUMPY_DTYPE)] * x.shape[0])
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py:138]: Can we include the mask processing in `_get_mask`?

File: art/attacks/evasion/hop_skip_jump.py (hunk 5 above)
Comment (beat-buesser) [reply art/attacks/evasion/hop_skip_jump.py:238]: This change is related to the previous comment on unifying function calls.

File: tests/attacks/test_hop_skip_jump.py [hunk 7]
```diff
@@ -81,6 +81,24 @@ def test_tensorflow_mnist(self):
         y_pred_adv = np.argmax(tfc.predict(x_test_adv), axis=1)
         self.assertTrue((target == y_pred_adv).any())
 
+        # Test the masking 1
```
Comment (beat-buesser) [inline tests/attacks/test_hop_skip_jump.py:84]: I think the tests check if the unmasked input does not get modified, which is good. Could you please also add a test to check if the masked input gets perturbed.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @minhitbk Thank you very much for adding support for masks to HopSkipJump. I have a few minor requests that could optimise the implementation.

File: art/attacks/evasion/hop_skip_jump.py (hunk 1 above)
Comment (minhitbk) [reply art/attacks/evasion/hop_skip_jump.py:156]: Yes it is mandatory. If shapes of x and mask are different, then broadcasting will give error. For example x shape is (10, 32, 32, 3) and mask shape is (10, 32, 3), the vectorised version will throw error.

File: art/attacks/evasion/hop_skip_jump.py (hunk 2 above)
Comment (minhitbk) [reply art/attacks/evasion/hop_skip_jump.py:259]: Solved.

File: art/attacks/evasion/hop_skip_jump.py (hunk 3 above)
Comment (minhitbk) [reply art/attacks/evasion/hop_skip_jump.py]: Solved.

File: art/attacks/evasion/hop_skip_jump.py (hunk 4 above)
Comment (minhitbk) [reply art/attacks/evasion/hop_skip_jump.py]: Solved.

File: art/attacks/evasion/hop_skip_jump.py (hunk 6 above)
Comment (minhitbk) [reply art/attacks/evasion/hop_skip_jump.py]: Solved.

File: tests/attacks/test_hop_skip_jump.py (hunk 7 above)
Comment (minhitbk) [reply tests/attacks/test_hop_skip_jump.py:84]: Solved.

**Comment by minhitbk** [comment]
> Hi @minhitbk Thank you very much for adding support for masks to HopSkipJump. I have a few minor requests that could optimise the implementation.

Solved according to your reviews. Thank you @beat-buesser !

File: art/attacks/evasion/hop_skip_jump.py [hunk 8]
```diff
@@ -158,25 +175,54 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+                    x_adv[ind] = self._perturb(
+                        x=val,
+                        y=y[ind],
+                        y_p=preds[ind],
+                        init_pred=init_preds[ind],
+                        adv_init=x_adv_init[ind],
+                        mask=None,
+                        clip_min=clip_min,
+                        clip_max=clip_max,
+                    )
```
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py]:
```suggestion
                x_adv[ind] = self._perturb(
                    x=val,
                    y=y[ind],
                    y_p=preds[ind],
                    init_pred=init_preds[ind],
                    adv_init=x_adv_init[ind],
                    mask=mask[ind],
                    clip_min=clip_min,
                    clip_max=clip_max,
                )
```

File: art/attacks/evasion/hop_skip_jump.py (hunk 4 above)
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py]:
```suggestion
                x_adv[ind] = self._perturb(
                    x=val,
                    y=-1,
                    y_p=preds[ind],
                    init_pred=init_preds[ind],
                    adv_init=x_adv_init[ind],
                    mask=mask[ind],
                    clip_min=clip_min,
                    clip_max=clip_max,
                )
```

File: art/attacks/evasion/hop_skip_jump.py (hunk 2 above)
Comment (beat-buesser) [inline art/attacks/evasion/hop_skip_jump.py]:
```suggestion
        if mask is None:
            mask = [None] * x.shape[0]
        else:
            if classifier_mixin:
                # Ensure the mask is broadcastable
                if len(mask.shape) > len(x.shape) or mask.shape != x.shape[-len(mask.shape) :]:
                    raise ValueError("Mask shape must be broadcastable to input shape.")

            else:
                raise ValueError("Mask is only supported for classification.")
```

**Review by beat-buesser — APPROVED** [review]
